sandbox/gpio/common.go

```diff
@@ -0,0 +1,276 @@
+type kernelChipInfo struct {
+	name, label [32]byte
+	lines       uint32
+}
+
+const _GPIO_GET_CHIPINFO_IOCTL uintptr = 0x8044b401
+
+var unixSyscall = unix.Syscall
+
+var ioctlGetChipInfo = func(path string) (*kernelChipInfo, error) {
+	f, err := os.OpenFile(path, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	conn, err := f.SyscallConn()
+	if err != nil {
+		return nil, err
+	}
+
+	kci := new(kernelChipInfo)
+	var errno syscall.Errno
+	err = conn.Control(func(fd uintptr) {
+		_, _, errno = unixSyscall(unix.SYS_IOCTL, f.Fd(), _GPIO_GET_CHIPINFO_IOCTL, uintptr(unsafe.Pointer(kci)))
+	})
+	if errno != 0 {
+		return nil, errno
+	}
+	return kci, err
+}
+
+var chardevChipInfo = func(path string) (*ChardevChip, error) {
+	kci, err := ioctlGetChipInfo(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read gpio chip info from %q: %w", path, err)
+	}
+
+	chip := &ChardevChip{
+		Path:     path,
+		Name:     string(bytes.TrimRight(kci.name[:], "\x00")),
+		Label:    string(bytes.TrimRight(kci.label[:], "\x00")),
+		NumLines: uint(kci.lines),
+	}
+	return chip, nil
+}
+
+const (
+	aggregatorLockPath         = "/sys/bus/platform/drivers/gpio-aggregator"
+	aggregatorNewDevicePath    = "/sys/bus/platform/drivers/gpio-aggregator/new_device"
+	aggregatorDeleteDevicePath = "/sys/bus/platform/drivers/gpio-aggregator/delete_device"
+	ephemeralUdevRulesDir      = "/run/udev/rules.d"
+)
+
+var lockAggregator = func() (unlocker func(), err error) {
+	flock, err := osutil.OpenExistingLockForReading(filepath.Join(dirs.GlobalRootDir, aggregatorLockPath))
+	if err != nil {
+		return nil, err
+	}
+	if err := flock.Lock(); err != nil {
+		return nil, err
+	}
+	return func() {
+		// we don't care about the error, this is best-effort.
+		_ = flock.Close()
+	}, nil
+}
+
+var aggregatorCreationTimeout = 120 * time.Second
+
+func addAggregatedChip(ctx context.Context, sourceChip *ChardevChip, lines strutil.Range) (chip *ChardevChip, err error) {
+	// synchronize gpio helpers' access to the aggregator interface
+	unlocker, err := lockAggregator()
+	if err != nil {
+		return nil, err
+	}
+	defer unlocker()
+
+	watcher, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	defer watcher.Close()
+
+	err = watcher.AddWatch(dirs.DevDir, inotify.InCreate)
+	if err != nil {
+		return nil, err
+	}
+
+	// <label> <lines>
+	cmd := fmt.Sprintf("%s %s", sourceChip.Label, lines.String())
+	if err := os.WriteFile(filepath.Join(dirs.GlobalRootDir, aggregatorNewDevicePath), []byte(cmd), 0644); err != nil {
+		return nil, err
+	}
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, aggregatorCreationTimeout)
+	defer cancel()
+	for {
+		select {
+		case event := <-watcher.Event:
+			path := event.Name
+			// check prefix /dev/gpiochipX
+			if strings.HasPrefix(path, filepath.Join(dirs.DevDir, "gpiochip")) {
+				return chardevChipInfo(path)
+			}
+		case <-ctxWithTimeout.Done():
+			return nil, ctxWithTimeout.Err()
```

Comment:
maybe a more meaningful error in place of a generic context timeout:
```suggestion
			return nil, errors.New("timeout waiting for aggregator device to appear")
```

Reply:
Good point, I used a separate timeout channel to avoid masking other context cancellation errors from parent context.